src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -36,6 +36,18 @@ object Literal {
   }
 }
 
+object NumericConstant {
+  def apply(x: Double, t: Type): IR = {
```

Comment:
This can be implemented in IR (and it will get optimized) as: `ir.Cast(I64(x), t)`.

Reply:
oh, nice! F64 though

Reply:
@cseed I changed the code to use `Cast`, but am getting an error in the TypeCheck. `assert(Casts.valid(v.typ, typ))` because the types are not in the Map because only the unrequired types are in the Map. To fix that, I changed the lookup code to this:

```
  def get(from: Type, to: Type): Code[_] => Code[_] =
    casts(from.setRequired(false) -> to.setRequired(false)).asInstanceOf[Code[_] => Code[_]]

  def valid(from: Type, to: Type): Boolean =
    casts.contains(from.setRequired(false) -> to.setRequired(false))
```

but then am getting a different assertion error in TypeCheck `x@Ref(name, _)` because of requiredness differences in the PL and AD fields.

Could you please verify the code I added should work? If so, I think I should revert my last commit and work on fixing the cast code in a different PR.

Reply:
@catoverdrive showed me how to fix this.